### PR TITLE
feat: testing games with artificial network conditions

### PR DESCRIPTION
**Pull request by pdeschain**
**Select the type of change:**

- [ ] Small Changes - Typos, formatting, slight revisions
- [x] New Content - New features, sections, pages, tutorials
- [ ] Site and Tools - Updates, maintenance, and new packages for the site and Docusaurus

**Purpose of the Pull Request:**
This PR adds a guide covering the available workflows for testing multiplayer games locally with artificial network conditions and some general advice and tricks that can be used with the available tools.

To-do:
 - [ ] Add ParrelSync video and send it to Brian with artificial network conditions  (Simulator Tools are causing trouble - talking to code owners)

**Issue Number:** 
MTT-978

### CODE REVIEW

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.docs?pullRequest=300) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.docs?pullRequest=300) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: dependabot[bot]<br/>:x: lorikrell<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.multiplayer.docs?pullRequest=300) it.</sub>

**Review by LukeStampfli — APPROVED** [review]
Overall this is great. Explains everything which is important around the topic and I think even beginners should be able to understand it.

I was kind of missing a sentence along the line of "While artificial latency is great for simulating network conditions during development it will not accurately emulate real world conditions. We recommend to test your game frequently on the targeted platforms and real live networking conditions.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:36]: mmm so it might be worth it to have suggestions for other platforms too. Or at least mention that most platforms have their own way to simulate artificial latency. iOS has dev tools for this, I'm sure android and consoles do too. 
It could be an addition later too, once we actually get to test this ourselves.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:45]: That part I'm not sure about. Users won't switch to a better internet to play a different game genre. They'll have their home internet and that's it, the lag you use to test shouldn't depend on the game type.
If I'm at home and I play an FPS or a card game, in both cases, my ping will be the same.
Per platform lag, that I agree. In past projects, I've seen a clear difference in our live analytics between mobile vs desktop/consoles (and even between iOS and Android).
So I would test with higher lag if I want to target mobile for example, since there's a higher chances live users will have higher latencies.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:47]: for the following few lines, platform will influence base lag too, not just chaos

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:55]:
```suggestion
Adding jitter to the base delay value adds a layer of chaotic unreliability that would make our peers behave in a more natural way, allowing us to tweak our interpolation, buffering and other techniques to compensate for these instabilities.
```

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:58]:
```suggestion
Adding packet loss, apart from introducing even more effective delay to our system could also wreak havoc on our unreliable messages, thus allowing us to explore if we need more defensive logic surrounding our unreliable messages or if we should opt for a reliable message instead.

Testing Boss Room on desktop, we've used artificial lag between 80ms and 120ms with 5%-10% packet loss.

### Different network conditions for different peers
```

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:38]:
```suggestion
## How much lag/packet loss/jitter should we use? 
```

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:74]:
```suggestion
[Unity Transport for Necode for GameObjects](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/tree/develop/com.unity.netcode.adapter.utp) comes with it's own tool for simulating artificial adverse network conditions - the Simulator Tools window.
```

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:127]:
```suggestion
 - They are system-wide, thus all the local instances of our game would run under the same network conditions.
```

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:153]: nothing on dnctl?

**Comment by SamuelBellomo** [comment]
+1 for Luke's comment on testing in real world

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:40]: nit: This general advice section could benefit from a bit of summarizing? @Briancoughlin @s-omeilia-unity what do you guys think?

**Comment by SamuelBellomo** [comment]
Added some comments, cool piece, very complete.
We'll be able to add more details for mobile and consoles too. I added MTT-1301 for mobile and MTT-1302 for mobile

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:36]: Yeah I thought about mobile platforms and even consoles, but until we can test it - feels weird to recommend solutions until we test them ourselves. I can add a general statement that be on the lookout for platform-specific tools if you are working on a platform other than Win / Mac.

Comment (SamuelBellomo) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:36]: good idea
For testing, we'll need to look into dev consoles very soon.
For mobile, in theory we should already have what it takes. I have iOS devices and @fernando-cortez has Android ones. I can reserve some time next week to test this if you want

Comment (chrispope) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:36]: I would expect we cover recommendations for the platforms we say we support. I understand the difficulties for consoles while we don't have dev kits. I suggest reaching out to the console support teams lead by Jason Page for knowledge and guidance.

**Comment by Briancoughlin** [comment]
We can update as we get kit as well.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (Briancoughlin) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:36]: So it's important that we remember that this document is live, we can add to it. I don't want to hold up its release. My question here is would this be releasable with its current content or would it need more content on the other platforms for launch.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (Briancoughlin) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:45]: Em, im not sure, I switched my internet provide to get better internet. im pretty sure a lot of gamers when they move house check the provider speeds(I know I do). Users will go to the game that requires the best and work from there so in your case I would work out what i need for the FLS and switch to that, knowing the other games will work fine on it.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (chrispope) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:36]: As I said, at the least the platforms that NGO supports.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (chrispope) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:45]: I'd expect a games studio to have a target (or idea) for lag tolerance across all the targeted platforms and test against that. i.e. just getting a handle on how unplayable the game could be on x platform with severe lag. 
As developers, we have zero control over the player's internet speeds and internet hardware setup. 

It might not be good for business to say - _please upgrade your internet to Starlink for optimal gameplay experiences_

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (chrispope) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:51]: Might be worth calling out UTPs Device Connection feature here: https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/blob/rfc/device-reconnection/text/0000-device-reconnection.md

**Comment by Briancoughlin** [comment]
@pdeschain  would it be possible to get me the ParallelSync as a MP4 instead of a GIf please?

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (Briancoughlin) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:45]: I agree we can't say upgrade your internet, but we already have a section that outlines examples of how lag impacts different game types, this section seems to fit in with that general guidance. So explaining that if your designing a game that is turn based  lag might not be as impactful as it is on a FPS so testing with artificially high lag in the latter would be more useful than the former. So for instance Overwatch mIn specs states you need a broadband connection. so dialup wont cut it.

https://eu.battle.net/support/en/article/65159

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:36]: Reached out to Jason Page regarding touching on the tools for consoles.
We had a small discussion with Sam and the current idea is:
 - document Unity's stuff (UTP + Parrelsync for now)
 - touch on some tools that are easy to touch on (osx and windows stuff)
 - mention the rest (NetworkLinkConditioner on iOS, console stuff, Android and so forth).

Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:36]: we want to avoid documenting the tools that other companies develop too heavily because it's a pain to keep these up to date. We would let people know that tools for their platforms exist and refer them to the docs.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:45]: Oh, so this bit is not about the user and their internet - it's about answering a question "what is the minimal latency I should even be testing with, and what is the typical latency under which other games in the same genre tend to fare well". It's ta gauge reasonable expectations for your own code - under what conditions should it perform adequately.

Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:45]: As in, the team needs to define that "our game should be playable within 0 to 150ms lag with no perceptible degradation of experience, 150-200 ms is bad, but it could still work, everything else we don't care about"

Comment (chrispope) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:45]: Yeah, I think we are eluding to the same topic, so maybe the feedback here is go a bit wider with information, as it could be misinterpreted or lacking details?

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (chrispope) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:36]: Yes, I meant to elaborate further; I certainly didn't mean we document the individual tools Unity are not responsible for developing but I wanted to see tools called out on a per platform basis. If we can link to them as well, great, otherwise just bring awareness is suitable.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:155]: Will add a section on DummyNet for sure.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md]: Rephrased this section a bit.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:36]: Overwatch fails elegantly above 300ms. In the sense where you can still play the game, but server rewind will stop helping you after 300ms and so you'll start missing your shots more. This is a choice Blizzard made. Other FPS devs will not put such a restriction and will allow 500ms and even 1000ms latency.
In past projects, our QA team would test with 100ms latency, 200ms, 500ms and 1000ms and give a report on playability. It'd then be up to designers/leads to determine if the issues found were acceptable or not, if they were ship blockers or not.
Determining which network conditions to test with is NOT about what type of game you have. It's about what your users will encounter while playing your game, which will vary by region and by platform, not by genre.
Saying an FPS will require lower latencies doesn't make sense --> tell that to your roommate trying to download game of thrones while you play. We have lag hiding techniques for this, we have prediction, we have server rewind, we have interpolation/buffering to hide that jitter, but that 200ms of lag won't disappear, it's completely our of your control as a player and as a dev. It's up to the dev to make sure it doesn't show. And to do that, you need to test with realistic network conditions and use the right techniques to hide it successfully.

Now determining what bugs are ACCEPTABLE after you tests latency, that's dependent on game genre. If I have a turn by turn card game and I don't get an update after 500ms due to latency, I wouldn't mind. Or if my tank is moving slow because I'm not predicting it, I wouldn't mind (I'd call this a feature and say it's due to the big thing's intertia... :P ) But you should still test at 100ms, 200ms, 500ms, 1000ms, to make sure you don't have weird race conditions, since your users WILL have those latencies.

Again, test values for artificial network conditions are orthogonal to game genre.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md]: Overwatch fails elegantly above 300ms. In the sense where you can still play the game, but server rewind will stop helping you after 300ms and so you'll start missing your shots more. This is a choice Blizzard made. Other FPS devs will not put such a restriction and will allow 500ms and even 1000ms latency.
In past projects, our QA team would test with 100ms latency, 200ms, 500ms and 1000ms and give a report on playability. It'd then be up to designers/leads to determine if the issues found were acceptable or not, if they were ship blockers or not.
Determining which network conditions to test with is NOT about what type of game you have. It's about what your users will encounter while playing your game, which will vary by region and by platform, not by genre.
Saying an FPS will require lower latencies doesn't make sense --> tell that to your roommate trying to download game of thrones while you play. We have lag hiding techniques for this, we have prediction, we have server rewind, we have interpolation/buffering to hide that jitter, but that 200ms of lag won't disappear, it's completely our of your control as a player and as a dev. It's up to the dev to make sure it doesn't show. And to do that, you need to test with realistic network conditions and use the right techniques to hide it successfully.

Now determining what bugs are ACCEPTABLE after you tests latency, that's dependent on game genre. If I have a turn by turn card game and I don't get an update after 500ms due to latency, I wouldn't mind. Or if my tank is moving slow because I'm not predicting it, I wouldn't mind (I'd call this a feature and say it's due to the big thing's intertia... :P ) But you should still test at 100ms, 200ms, 500ms, 1000ms, to make sure you don't have weird race conditions, since your users WILL have those latencies.

Again, test values for artificial network conditions are orthogonal to game genre.

Comment (SamuelBellomo) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md]: I'll add
*what's acceptable is dependant on game genre AND on your game design team and leads. Market analysts could show you should target users from countries with higher latencies for example. It's up to you to determine what's acceptable, there's no single rule for this.
but you should still test different latencies!!

Comment (SamuelBellomo) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md]: Brian your link on Overwatch requirements specifies bandwidth requirements, not latency requirements. They are very different. You can have good bandwidth and very bad latency. Notice how they don't specify required ping to play

Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md]: Thank you Sam, you're 100% right. I've reworded the segment to reflect ^^.

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:55]:
```suggestion
To test Boss Room, we used around 100ms-150ms for desktop platforms and around 200ms-300ms for mobile platforms. These are just what we used, the actual values you should test with will depend on your target platform and region.
```

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:59]:
```suggestion
It's valuable to test at 100ms, 200ms, 500ms, even 1000ms and potentially higher to make sure that we don't have weird race conditions that present at different latency ranges. The users in the wild WILL have those latencies.
```

Comment (chrispope) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md]: I'm being a bit picky here, but I would prefer we use the term "players" vs "users".

Comment (SamuelBellomo) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md]: @pdeschain in case you dont see this in a resolve conversation

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:77]: mentioned the numbers we use elsewhere in the text

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:14]: nit: I think apple now writes it with a different capitalization (macOS) https://www.apple.com/ca/macos/big-sur/

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:81]: not just games, all programs including your web browsers. It might be worth it point out to not forget to disable it once you're done, else your computer will feel slow :P

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (SamuelBellomo) [inline docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:163]: I see "recommend" in multiple places in this doc. @chrispope @Briancoughlin should we really endorse these?
I'm thinking more "we've used these in our own development".

Comment (chrispope) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:163]: I would suggest we author this in a way that brings awareness of third-party tools, but we shouldn't endorse or recommend any in particular. 

No harm in saying this is what we have used during our development.

I would prefer we come across impartial and users can make their own assessments on the tools which best suit them.

Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md]: agreed, rephrased

File: docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md
Comment (pdeschain) [reply docs/tutorials/local_iteration_series/testing_with_artificial_conditions.md:81]: agreed, added!

**Pull request merged**
